src/Hamiltonians/abstract.jl

```diff
@@ -0,0 +1,257 @@
+###
+### This file contains abstract types, interfaces and traits.
+###
+"""
+    AbstractHamiltonian{T}
+
+Supertype that provides an interface for linear operators over a linear space with scalar
+type `T` that are suitable for FCIQMC. Indexing is done with addresses (typically not
+integers) from an address space that may be large (and will not need to be completely
+generated).
+
+`AbstractHamiltonian` instances operate on vectors of type [`AbstractDVec`](@ref) from the
+module `DictVectors` and work well with addresses of type [`AbstractFockAddress`](@ref) from
+the module `BitStringAddresses`. The type works well with the external package
+[KrylovKit.jl](https://github.com/Jutho/KrylovKit.jl).
+
+# Methods
+
+Provides:
+
+* [`offdiagonals`](@ref): iterator over reachable off-diagonal matrix elements
+* [`generateRandHop`](@ref): function to generate random off-diagonal matrix element
+* [`dimension`](@ref): get the dimension of the address space.
+* `H[address1, address2]`: indexing with `getindex()` - mostly for testing purposes
+* `*(H, v)`: deterministic matrix-vector multiply.
+* `H(v)`: equivalent to `LO * v`.
+* `mul!(w, H, v)`: mutating matrix-vector multiply.
+* [`dot(x, H, v)`](@ref): compute `x⋅(LO*v)` minimizing allocations.
+
+Methods that need to be implemented:
+
+* [`num_offdiagonals(::AbstractHamiltonian, address)`](@ref)
+* [`get_offdiagonal(::AbstractHamiltonian, address, chosen::Integer)`](@ref)
+* [`diagonal_element(::AbstractHamiltonian, address)`](@ref)
+* [`starting_address(::AbstractHamiltonian)`](@ref)
+
+Optional methods to implement:
+
+* [`Hamiltonians.LOStructure(::Type{typeof(lo)})`](@ref)
+* [`dimension(::Type{T}, ::AbstractHamiltonian)`](@ref)
+* [`offdiagonals(::AbstractHamiltonian, ::AbstractFockAddress)`](@ref)
+* [`momentum(::AbstractHamiltonian)`](@ref)
+"""
+abstract type AbstractHamiltonian{T} end
+
+Base.eltype(::AbstractHamiltonian{T}) where {T} = T
+
+(h::AbstractHamiltonian)(v) = h * v
+(h::AbstractHamiltonian)(w, v) = mul!(w, h, v)
+
+BitStringAddresses.num_modes(h::AbstractHamiltonian) = num_modes(starting_address(h))
+
+"""
+    diagonal_element(ham, add)
+
+Compute the diagonal matrix element of the linear operator `ham` at
+address `add`.
+
+# Example
+
+```jldoctest
+julia> addr = BoseFS((3, 2, 1));
+
+julia> H = HubbardMom1D(addr);
+
+julia> diagonal_element(H, addr)
+8.666666666666664
+```
+"""
+diagonal_element
+
+"""
+    num_offdiagonals(ham, add)
+
+Compute the number of number of reachable configurations from address `add`.
+
+# Example
+
+```jldoctest
+julia> addr = BoseFS((3, 2, 1));
+
+julia> H = HubbardMom1D(addr);
+
+julia> num_offdiagonals(H, addr)
+10
+```
+"""
+num_offdiagonals
+
+"""
+    newadd, me = get_offdiagonal(ham, add, chosen)
+
+Compute matrix element of `hamiltonian` and new address of a single get_offdiagonal from
+address `add` with integer index `chosen`.
```

Comment:
Looks like copy and past went wrong. I suggest:
Compute matrix element `me` of  Hamiltonian `ham` and new address `newadd` of a single (off-diagonal) matrix element in the same column from the address `add` with integer index `chosen`.

Reply:
Oops, I used `sed` for this rename and missed this replacement. Fixed.